File: app/src/main/java/com/petnbu/petnbu/repo/UserRepository.java
package com.petnbu.petnbu.repo;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.petnbu.petnbu.AppExecutors;
import com.petnbu.petnbu.api.ApiResponse;
import com.petnbu.petnbu.api.WebService;
import com.petnbu.petnbu.db.PetDb;
import com.petnbu.petnbu.db.UserDao;
import com.petnbu.petnbu.model.Resource;
import com.petnbu.petnbu.model.UserEntity;

import javax.inject.Inject;
import javax.inject.Singleton;

import timber.log.Timber;

@Singleton
public class UserRepository {

    private final PetDb mPetDb;

    private final UserDao mUserDao;

    private final AppExecutors mAppExecutors;

    private final WebService mWebService;

    @Inject
    public UserRepository(PetDb petDb, UserDao userDao, AppExecutors appExecutors, WebService webService) {
        mPetDb = petDb;
        mUserDao = userDao;
        mAppExecutors = appExecutors;
        mWebService = webService;
    }

    public LiveData<Resource<UserEntity>> getUserById(@NonNull String id){
        Timber.i("getUserById: %s", id);
        return new NetworkBoundResource<UserEntity, UserEntity>(mAppExecutors){
            @Override
            protected void saveCallResult(@NonNull UserEntity item) {
                mUserDao.insert(item);
            }

            @Override
            protected boolean shouldFetch(@Nullable UserEntity data) {
                return true;
            }

            @NonNull
            @Override
            protected LiveData<UserEntity> loadFromDb() {
                return mUserDao.findLiveUserById(id);
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<UserEntity>> createCall() {
                return mWebService.getUser(id);
            }

            @Override
            protected void deleteDataFromDb(UserEntity body) {

            }
        }.asLiveData();
    }
}

File: app/src/main/java/com/petnbu/petnbu/repo/UserRepository.kt
package com.petnbu.petnbu.repo

import android.arch.lifecycle.LiveData
import com.petnbu.petnbu.AppExecutors
import com.petnbu.petnbu.api.ApiResponse
import com.petnbu.petnbu.api.WebService
import com.petnbu.petnbu.db.PetDb
import com.petnbu.petnbu.model.Resource
import com.petnbu.petnbu.model.UserEntity
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class UserRepository @Inject
constructor(private val mPetDb: PetDb, private val mAppExecutors: AppExecutors, private val mWebService: WebService) {

    fun getUserById(id: String): LiveData<Resource<UserEntity>> {
        Timber.i("getUserById: %s", id)
        return object : NetworkBoundResource<UserEntity, UserEntity>(mAppExecutors) {
            override fun saveCallResult(item: UserEntity) = mPetDb.userDao().insert(item)

            override fun shouldFetch(data: UserEntity?): Boolean = true

            override fun loadFromDb(): LiveData<UserEntity> = mPetDb.userDao().findLiveUserById(id)

            override fun createCall(): LiveData<ApiResponse<UserEntity>> = mWebService.getUser(id)

            override fun deleteDataFromDb(body: UserEntity?) {}
        }.asLiveData()
    }
}
